File: misc/twosum/TwoSum.py
class TwoSum:
    def twoSum(self, nums, target):
        # Create a dictionary to store the number and its index
        num_map = {}

        # Iterate over the list of numbers
        for i, num in enumerate(nums):
            complement = target - num  # Find the complement

            # Check if the complement is already in the dictionary
            if complement in num_map:
                # Return the indices of the complement and the current number
                return [num_map[complement], i]

            # Store the current number and its index in the dictionary
            num_map[num] = i

        # Return an empty list if no solution is found
        return []

# Sample usage
if __name__ == "__main__":
    solution = TwoSum()

    # Test case 1
    nums = [2, 7, 11, 15]
    target = 9
    print(solution.twoSum(nums, target))  # Output: [0, 1]

    # Test case 2
    nums = [3, 2, 4]
    target = 6
    print(solution.twoSum(nums, target))  # Output: [1, 2]

    # Test case 3
    nums = [3, 3]
    target = 6
    print(solution.twoSum(nums, target))  # Output: [0, 1]

File: misc/twosum/TwoSum.java
package twosum;

import java.util.HashMap;
import java.util.Map;

class TwoSum {
    public int[] twoSum(int[] nums, int target) {
        // Create a map to store the number and its index
        Map<Integer, Integer> map = new HashMap<>();

        // Iterate over the array
        for (int i = 0; i < nums.length; i++) {
            int second = target - nums[i];  // Find the complement

            // Check if the complement is already in the map
            if (map.containsKey(second)) {
                // Return the indices of the complement and the current number
                return new int[] { map.get(second), i };
            }

            // Store the current number and its index in the map
            map.put(nums[i], i);
        }

        // Return an empty array if no solution is found
        return new int[] {};
    }

    // Sample usage
    public static void main(String[] args) {
        TwoSum solution = new TwoSum();

        // Test case 1
        int[] nums1 = {2, 7, 11, 15};
        int target1 = 9;
        int[] result1 = solution.twoSum(nums1, target1);
        System.out.println("Indices: [" + result1[0] + ", " + result1[1] + "]");  // Output: Indices: [0, 1]

        // Test case 2
        int[] nums2 = {3, 2, 4};
        int target2 = 6;
        int[] result2 = solution.twoSum(nums2, target2);
        System.out.println("Indices: [" + result2[0] + ", " + result2[1] + "]");  // Output: Indices: [1, 2]

        // Test case 3
        int[] nums3 = {3, 3};
        int target3 = 6;
        int[] result3 = solution.twoSum(nums3, target3);
        System.out.println("Indices: [" + result3[0] + ", " + result3[1] + "]");  // Output: Indices: [0, 1]
    }
}
